fetch2/src/main/java/com/tonyodev/fetch2/downloader/FileDownloader.kt

```diff
@@ -33,4 +35,35 @@ interface FileDownloader : Runnable {
 
     }
 
-}
\ No newline at end of file
+}
+
+abstract class AbsFileDownloader(
+    protected val initialDownload: Download
+) : FileDownloader {
+
+    fun checkDownloadSpaceSafely(response: Downloader.Response): Boolean {
+        return response.contentLength.let { len ->
+            if (len < 0) {
+                return@let true
+            }
+            initialDownload.file.let {
+                val spaceLeft = when {
+                    it.startsWith("/data/data") -> {
+                        Environment.getDataDirectory().freeSpace
+                    }
+                    Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED -> {
+                        Environment.getExternalStorageDirectory().freeSpace
+                    }
+                    else -> 0L
+                }
+                spaceLeft > len.times(1.5)
```

Comment:
Why 1.5?

Reply:
For safe-space left only

Reply:
Not sure if it'll really work well for big files